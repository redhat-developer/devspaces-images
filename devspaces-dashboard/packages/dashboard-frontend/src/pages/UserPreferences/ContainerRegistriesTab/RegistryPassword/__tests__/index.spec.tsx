/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { RegistryPasswordFormGroup } from '..';

describe('Registry Password Input', () => {
  const mockOnChange = jest.fn();

  function getComponent(password: string): React.ReactElement {
    return (
      <Form>
        <RegistryPasswordFormGroup password={password} onChange={mockOnChange} />
      </Form>
    );
  }

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should correctly render the component', () => {
    const component = getComponent('*ttt#');
    render(component);

    const input = screen.queryByTestId('registry-password-input');
    expect(input).toBeTruthy();

    const json = renderer.create(component).toJSON();

    expect(json).toMatchSnapshot();
  });

  it('should show default placeholder', () => {
    const component = getComponent('');
    render(component);

    const placeholder = screen.queryByPlaceholderText('Enter a password');
    expect(placeholder).toBeTruthy();
  });

  it('should correctly re-render the component', () => {
    const component = getComponent('*test');
    const { rerender } = render(component);

    const input = screen.getByTestId('registry-password-input');
    expect(input).toHaveValue('*test');

    rerender(
      <Form>
        <RegistryPasswordFormGroup password={'*testreg#'} onChange={mockOnChange} />
      </Form>,
    );

    expect(input).toHaveValue('*testreg#');
  });

  it('should fire onChange event', () => {
    const component = getComponent('*te');
    render(component);

    const input = screen.getByTestId('registry-password-input');

    expect(input).toHaveValue('*te');
    expect(mockOnChange).not.toHaveBeenCalled();

    userEvent.type(input, 'st');

    expect(mockOnChange).toHaveBeenCalledWith('*tes', 'success');
    expect(mockOnChange).toHaveBeenCalledWith('*test', 'success');
  });

  describe('validation', () => {
    it('should handle empty value', () => {
      const component = getComponent('https://testreg.com/test1');
      render(component);

      const input = screen.getByTestId('registry-password-input');

      userEvent.clear(input);

      const label = screen.queryByText('A value is required.');

      expect(label).toBeTruthy();
      expect(input).toBeInvalid();
    });

    it('should handle maximum value length', () => {
      const component = getComponent('https://testreg.com/test1');
      render(component);

      const input = screen.getByTestId('registry-password-input');

      const message = 'The password is too long. The maximum length is 10000 characters.';
      const allowedPassword = 'a'.repeat(10000);
      let label: HTMLElement | null;

      userEvent.clear(input);
      userEvent.type(input, allowedPassword);

      label = screen.queryByText(message);
      expect(label).toBeFalsy();

      const disallowedPassword = 'a'.repeat(10001);

      userEvent.clear(input);
      userEvent.type(input, disallowedPassword);

      label = screen.queryByText(message);
      expect(label).toBeTruthy();
      expect(input).toBeInvalid();
    });
  });
});
