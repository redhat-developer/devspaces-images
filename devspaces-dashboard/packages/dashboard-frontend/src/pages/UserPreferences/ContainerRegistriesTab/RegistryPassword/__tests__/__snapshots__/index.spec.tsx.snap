// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registry Password Input should correctly render the component 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
>
  <div
    className="pf-c-form__group"
    style={
      Object {
        "gridTemplateColumns": "80px",
        "minHeight": "65px",
      }
    }
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="id-password-helper"
      >
        <span
          className="pf-c-form__label-text"
        >
          Password
        </span>
        <span
          aria-hidden="true"
          className="pf-c-form__label-required"
        >
           
          *
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <span
        className="pf-c-input-group__text"
      >
        <input
          aria-invalid={false}
          aria-label="Password input"
          className="pf-c-form-control"
          data-testid="registry-password-input"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Enter a password"
          readOnly={false}
          required={false}
          type="password"
          value="*ttt#"
        />
        <button
          aria-disabled={false}
          aria-label="show"
          className="pf-c-button pf-m-control"
          data-ouia-component-id="OUIA-Generated-Button-control-2"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe={true}
          disabled={false}
          onClick={[Function]}
          role={null}
          type="button"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 640 512"
            width="1em"
          >
            <path
              d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
            />
          </svg>
        </button>
      </span>
      
    </div>
  </div>
</form>
`;
